x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/query/DenseVectorScriptDocValues.java

```diff
@@ -8,87 +8,84 @@
 
 package org.elasticsearch.xpack.vectors.query;
 
-import org.apache.lucene.index.BinaryDocValues;
 import org.apache.lucene.util.BytesRef;
-import org.elasticsearch.Version;
 import org.elasticsearch.index.fielddata.ScriptDocValues;
 import org.elasticsearch.script.field.Field;
-import org.elasticsearch.xpack.vectors.mapper.VectorEncoderDecoder;
 
-import java.io.IOException;
+public abstract class DenseVectorScriptDocValues extends ScriptDocValues<BytesRef> {
+    public static final String MISSING_VECTOR_FIELD_MESSAGE = "A document doesn't have a value for a vector field!";
 
-public class DenseVectorScriptDocValues extends ScriptDocValues<BytesRef> {
-
-    private final BinaryDocValues in;
-    private final Version indexVersion;
     private final int dims;
-    private final float[] vector;
-    private BytesRef value;
-
 
-    DenseVectorScriptDocValues(BinaryDocValues in, Version indexVersion, int dims) {
-        this.in = in;
-        this.indexVersion = indexVersion;
+    public DenseVectorScriptDocValues(int dims) {
         this.dims = dims;
-        this.vector = new float[dims];
-    }
-
-    @Override
-    public void setNextDocId(int docId) throws IOException {
-        if (in.advanceExact(docId)) {
-            value = in.binaryValue();
-        } else {
-            value = null;
-        }
     }
 
-    // package private access only for {@link ScoreScriptUtils}
-    BytesRef getEncodedValue() {
-        return value;
-    }
-
-    // package private access only for {@link ScoreScriptUtils}
-    int dims() {
+    public int dims() {
         return dims;
     }
 
-    @Override
-    public BytesRef get(int index) {
-        throw new UnsupportedOperationException("accessing a vector field's value through 'get' or 'value' is not supported!" +
-            "Use 'vectorValue' or 'magnitude' instead!'");
-    }
-
     /**
      * Get dense vector's value as an array of floats
      */
-    public float[] getVectorValue() {
-        VectorEncoderDecoder.decodeDenseVector(value, vector);
-        return vector;
-    }
+    public abstract float[] getVectorValue();
 
     /**
      * Get dense vector's magnitude
      */
-    public float getMagnitude() {
-        return VectorEncoderDecoder.getMagnitude(indexVersion, value);
-    }
+    public abstract float getMagnitude();
 
-    @Override
-    public int size() {
-        if (value == null) {
-            return 0;
-        } else {
-            return 1;
-        }
-    }
+    public abstract double dotProduct(float[] queryVector);
+    public abstract double l1Norm(float[] queryVector);
+    public abstract double l2Norm(float[] queryVector);
```

Comment:
Why we did not include `cosineSimilarity` to the list of these functions?

Reply:
I thought about adding it, but I didn't see much benefit and the API contract turned out strange. Once per query, we make sure to normalize the query vector to unit length. So this method would have to mean "do cosine similarity with a query vector that's always unit length".